printer/printer.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2023 Terramate GmbH
+// SPDX-License-Identifier: MPL-2.0
+
+package printer
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/fatih/color"
+)
+
+var (
+	bold       = color.New(color.Bold).Sprint
+	boldYellow = color.New(color.Bold, color.FgYellow).Sprint
+	boldRed    = color.New(color.Bold, color.FgRed).Sprint
+	boldGreen  = color.New(color.Bold, color.FgGreen).Sprint
+)
+
+// Printer encapuslates an io.Writer and an error parser function
+type Printer struct {
+	w io.Writer
+	// errorParserFunc is a function that will be used to convert the provided
+	// error(s) to a human readable form
+	errorParserFunc func(error) []string
+}
+
+func defaultErrorParser(err error) []string {
+	return []string{err.Error()}
+}
+
+// NewPrinter creates a new Printer with a default error parser
+func NewPrinter(w io.Writer) *Printer {
+	return &Printer{w, defaultErrorParser}
+}
+
+// NewPrinterWithErrorParser creates a new printer that will use the provided
+// errorParserFunc to parse errors.
+func NewPrinterWithErrorParser(w io.Writer, errorParserFunc func(error) []string) *Printer {
+	return &Printer{w, errorParserFunc}
+}
+
+// Println prints a message to the io.Writer
+func (t *Printer) Println(msg string) {
+	fmt.Fprintln(t.w, msg)
+}
+
+// Warnln prints a message with a "Warning:" prefix. The prefix is printed in
+// the boldYellow style.
+func (t *Printer) Warnln(title string) {
+	fmt.Fprintln(t.w, boldYellow("Warning:"), bold(title))
+}
+
+// ErrorWithDetailsln prints an error with a title and the underlying error
+// e.g.:
+/* Error: parsing failed
+ * -> somefile.tm:8,3-7: terramate schema error: unrecognized attribute
+ */
+func (t *Printer) ErrorWithDetailsln(title string, err error) {
+	t.Errorln(title)
+	for _, errString := range t.errorParserFunc(err) {
+		fmt.Fprintln(t.w, boldRed("->"), errString)
+	}
```

Comment:
You can provide the error details without injecting an "error parser", just need to import Terramate "errors" package. Example:

```go
errs := errors.L()
switch e := err.(type) {
case *errors.Error:
	// an error can wrap a list of errors.
	errs = e.AsList()
case *errors.List:
	errs = e
default:
	errs = errors.L(err)
}
```
This way you get an `*errors.List` for any err.

Then you can do:
```go
for _, err := range errs.Errors() {
    fmt.Fprintln(t.w, boldRed("->"), err.Error())
}
```

Reply:
As discussed, I changed the implementation to use `errors.L(err).Errors()` to loop over the provided errors. Please see latest diff.